### ISSUE TITLE
Not all mappings are learnable

### ISSUE THREAD

**Issue by wxhenry** [alanbacellar/DWN#2]
When running DWN_Tutorial, I found that only the first LUTLayer is set to have learnable mapping, while the second LUTLayer have a fixed random mapping. I tried to set the second LUTLayer to have learnable mapping, but it turns out to perform worse than the fixed random mapping. I wonder if there is any reason for this phenomenon. Is it because Vanishing Gradient or other reasons?

**Comment by alanbacellar** [comment]
This behavior is mostly due to unstable gradients. When the argmax mapping changes during the forward pass, it can cause vastly different gradients during backpropagation, leading to noisy and worse convergence.  Using learnable mapping on the second layer may also cause overfitting. To stabilize training when using learnable mapping in the second layer, fine-tuning its gradient temperature parameter is necessary, but this significantly increases training time.

In our experiments, we observed that using learnable mappings in both LUT layers does not improve final accuracy compared to using learnable mappings only in the first layer. However, it requires much more fine-tuning effort to converge properly.

As noted in the README, we recommend using learnable mappings only in the first layer for more stable and efficient training.

**Comment by wxhenry** [comment]
Thanks for the detailed explanation! Really helped me understand the learnable mappings better. Appreciate it!
